In [119]:
import pandas as pd
import numpy as np

In [120]:
path = "E:\\data_analytics\\powerbi_projects\\Indian_crimes\\crime_dataset_india.csv"
df = pd.read_csv(path)
df.head()

,Report Number,Date Reported,Date of Occurrence,Time of Occurrence,City,Crime Code,Crime Description,Victim Age,Victim Gender,Weapon Used,Crime Domain,Police Deployed,Case Closed,Date Case Closed
0,1,02-01-2020 00:00,01-01-2020 00:00,01-01-2020 01:11,Ahmedabad,576,IDENTITY THEFT,16,M,Blunt Object,Violent Crime,13,No,NaN
1,2,01-01-2020 19:00,01-01-2020 01:00,01-01-2020 06:26,Chennai,128,HOMICIDE,37,M,Poison,Other Crime,9,No,NaN
2,3,02-01-2020 05:00,01-01-2020 02:00,01-01-2020 14:30,Ludhiana,271,KIDNAPPING,48,F,Blunt Object,Other Crime,15,No,NaN
3,4,01-01-2020 05:00,01-01-2020 03:00,01-01-2020 14:46,Pune,170,BURGLARY,49,F,Firearm,Other Crime,1,Yes,29-04-2020 05:00
4,5,01-01-2020 21:00,01-01-2020 04:00,01-01-2020 16:51,Pune,421,VANDALISM,30,F,Other,Other Crime,18,Yes,08-01-2020 21:00


In [121]:
df.shape

(40160, 14)

In [122]:
df.isnull().sum()

Report Number             0
Date Reported             0
Date of Occurrence        0
Time of Occurrence        0
City                      0
Crime Code                0
Crime Description         0
Victim Age                0
Victim Gender             0
Weapon Used            5790
Crime Domain              0
Police Deployed           0
Case Closed               0
Date Case Closed      20098
dtype: int64

In [123]:
df[['Victim Age','Police Deployed']].describe()

,Victim Age,Police Deployed
count,40160.00000,40160.000000
mean,44.49126,10.006250
std,20.22555,5.467951
min,10.00000,1.000000
25%,27.00000,5.000000
50%,44.00000,10.000000
75%,62.00000,15.000000
max,79.00000,19.000000


In [124]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40160 entries, 0 to 40159
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Report Number       40160 non-null  int64 
 1   Date Reported       40160 non-null  object
 2   Date of Occurrence  40160 non-null  object
 3   Time of Occurrence  40160 non-null  object
 4   City                40160 non-null  object
 5   Crime Code          40160 non-null  int64 
 6   Crime Description   40160 non-null  object
 7   Victim Age          40160 non-null  int64 
 8   Victim Gender       40160 non-null  object
 9   Weapon Used         34370 non-null  object
 10  Crime Domain        40160 non-null  object
 11  Police Deployed     40160 non-null  int64 
 12  Case Closed         40160 non-null  object
 13  Date Case Closed    20062 non-null  object
dtypes: int64(4), object(10)
memory usage: 4.3+ MB


In [125]:
df.duplicated().sum()

np.int64(0)

In [126]:
df['Weapon Used']=df['Weapon Used'].fillna('Unknown')

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40160 entries, 0 to 40159
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Report Number       40160 non-null  int64 
 1   Date Reported       40160 non-null  object
 2   Date of Occurrence  40160 non-null  object
 3   Time of Occurrence  40160 non-null  object
 4   City                40160 non-null  object
 5   Crime Code          40160 non-null  int64 
 6   Crime Description   40160 non-null  object
 7   Victim Age          40160 non-null  int64 
 8   Victim Gender       40160 non-null  object
 9   Weapon Used         40160 non-null  object
 10  Crime Domain        40160 non-null  object
 11  Police Deployed     40160 non-null  int64 
 12  Case Closed         40160 non-null  object
 13  Date Case Closed    20062 non-null  object
dtypes: int64(4), object(10)
memory usage: 4.3+ MB


In [128]:
df['Weapon Used'].value_counts()

Weapon Used
Knife           5835
Unknown         5790
Explosives      5751
Blunt Object    5737
Poison          5728
Other           5676
Firearm         5643
Name: count, dtype: int64

In [129]:
df['Weapon Used'].value_counts(normalize=True)*100

Weapon Used
Knife           14.529382
Unknown         14.417331
Explosives      14.320219
Blunt Object    14.285359
Poison          14.262948
Other           14.133466
Firearm         14.051295
Name: proportion, dtype: float64

These are numbers and percentage of reported crimes where specific weapons were used.

In [130]:
df['City'].value_counts().head(10)


City
Delhi        5400
Mumbai       4415
Bangalore    3588
Hyderabad    2881
Kolkata      2518
Chennai      2493
Pune         2212
Ahmedabad    1817
Jaipur       1479
Lucknow      1456
Name: count, dtype: int64

In [131]:
Cases_reported_percentage = df['City'].value_counts(normalize=True)*100
Cases_reported_percentage.head(10)

City
Delhi        13.446215
Mumbai       10.993526
Bangalore     8.934263
Hyderabad     7.173805
Kolkata       6.269920
Chennai       6.207669
Pune          5.507968
Ahmedabad     4.524402
Jaipur        3.682769
Lucknow       3.625498
Name: proportion, dtype: float64

These are number and percentage of crimes reported across Top 10 cites.

In [132]:
df['Crime Description'].value_counts()

Crime Description
BURGLARY               1980
VANDALISM              1975
FRAUD                  1965
DOMESTIC VIOLENCE      1932
FIREARM OFFENSE        1931
ROBBERY                1928
KIDNAPPING             1920
IDENTITY THEFT         1918
SEXUAL ASSAULT         1917
ASSAULT                1915
TRAFFIC VIOLATION      1915
PUBLIC INTOXICATION    1912
HOMICIDE               1909
CYBERCRIME             1899
ILLEGAL POSSESSION     1895
ARSON                  1894
DRUG OFFENSE           1890
EXTORTION              1873
COUNTERFEITING         1871
VEHICLE - STOLEN       1862
SHOPLIFTING            1859
Name: count, dtype: int64

In [133]:
df['Crime Description'].value_counts(normalize=True)*100

Crime Description
BURGLARY               4.930279
VANDALISM              4.917829
FRAUD                  4.892928
DOMESTIC VIOLENCE      4.810757
FIREARM OFFENSE        4.808267
ROBBERY                4.800797
KIDNAPPING             4.780876
IDENTITY THEFT         4.775896
SEXUAL ASSAULT         4.773406
ASSAULT                4.768426
TRAFFIC VIOLATION      4.768426
PUBLIC INTOXICATION    4.760956
HOMICIDE               4.753486
CYBERCRIME             4.728586
ILLEGAL POSSESSION     4.718625
ARSON                  4.716135
DRUG OFFENSE           4.706175
EXTORTION              4.663845
COUNTERFEITING         4.658865
VEHICLE - STOLEN       4.636454
SHOPLIFTING            4.628984
Name: proportion, dtype: float64

These are number and percentage of crimes reported across each crime descriptions.

In [134]:
df['Victim Gender'].value_counts()

Victim Gender
F    22423
M    13405
X     4332
Name: count, dtype: int64

In [135]:
df['Victim Gender'].value_counts(normalize=True)*100

Victim Gender
F    55.834163
M    33.378984
X    10.786853
Name: proportion, dtype: float64

Number and percentage of cases reported by victim gender: female, male, and transgender.

In [136]:
df['Crime Domain'].value_counts()

Crime Domain
Other Crime         22948
Violent Crime       11472
Fire Accident        3825
Traffic Fatality     1915
Name: count, dtype: int64

In [137]:
df['Crime Domain'].value_counts(normalize=True)*100

Crime Domain
Other Crime         57.141434
Violent Crime       28.565737
Fire Accident        9.524402
Traffic Fatality     4.768426
Name: proportion, dtype: float64

Number and percentage of cases repoted across each crime domain.

In [138]:
df['Case Closed'].value_counts()

Case Closed
No     20098
Yes    20062
Name: count, dtype: int64

In [139]:
df['Case Closed'].value_counts(normalize=True)*100

Case Closed
No     50.044821
Yes    49.955179
Name: proportion, dtype: float64

Number and percentage of cases still open or closed.

In [140]:
df['Date Reported']= pd.to_datetime(df['Date Reported'], format="%d-%m-%Y %H:%M")
df['Date Case Closed']= pd.to_datetime(df['Date Case Closed'], format="%d-%m-%Y %H:%M")

In [141]:
df['Days Took To Close']= (df['Date Case Closed']-df['Date Reported']).dt.days

In [142]:
df.head(10)

,Report Number,Date Reported,Date of Occurrence,Time of Occurrence,City,Crime Code,Crime Description,Victim Age,Victim Gender,Weapon Used,Crime Domain,Police Deployed,Case Closed,Date Case Closed,Days Took To Close
0,1,2020-01-02 00:00:00,01-01-2020 00:00,01-01-2020 01:11,Ahmedabad,576,IDENTITY THEFT,16,M,Blunt Object,Violent Crime,13,No,NaT,NaN
1,2,2020-01-01 19:00:00,01-01-2020 01:00,01-01-2020 06:26,Chennai,128,HOMICIDE,37,M,Poison,Other Crime,9,No,NaT,NaN
2,3,2020-01-02 05:00:00,01-01-2020 02:00,01-01-2020 14:30,Ludhiana,271,KIDNAPPING,48,F,Blunt Object,Other Crime,15,No,NaT,NaN
3,4,2020-01-01 05:00:00,01-01-2020 03:00,01-01-2020 14:46,Pune,170,BURGLARY,49,F,Firearm,Other Crime,1,Yes,2020-04-29 05:00:00,119.0
4,5,2020-01-01 21:00:00,01-01-2020 04:00,01-01-2020 16:51,Pune,421,VANDALISM,30,F,Other,Other Crime,18,Yes,2020-01-08 21:00:00,7.0
5,6,2020-01-02 03:00:00,01-01-2020 05:00,01-01-2020 17:09,Delhi,442,ASSAULT,16,M,Firearm,Violent Crime,18,Yes,2020-03-30 03:00:00,88.0
6,7,2020-01-01 16:00:00,01-01-2020 06:00,01-01-2020 14:08,Chennai,172,VEHICLE - STOLEN,64,F,Knife,Violent Crime,13,Yes,2020-03-24 16:00:00,83.0
7,8,2020-01-02 10:00:00,01-01-2020 07:00,02-01-2020 06:33,Chennai,169,COUNTERFEITING,78,X,Knife,Other Crime,8,No,NaT,NaN
8,9,2020-01-04 03:00:00,01-01-2020 08:00,02-01-2020 06:34,Mumbai,338,EXTORTION,41,X,Blunt Object,Other Crime,1,No,NaT,NaN
9,10,2020-01-03 07:00:00,01-01-2020 09:00,01-01-2020 17:50,Chennai,497,PUBLIC INTOXICATION,29,M,Knife,Other Crime,4,No,NaT,NaN


In [143]:
Avg_days_took_close = df['Days Took To Close'].mean()
Avg_Days_Took_to_Close = np.round(Avg_days_took_close, 2)
Avg_Days_Took_to_Close 

np.float64(88.06)

It is the average days took to close cases.

In [144]:
df.groupby('Crime Description')['Days Took To Close'].mean().sort_values(ascending=False)

Crime Description
HOMICIDE               453.056842
SEXUAL ASSAULT         447.768568
ROBBERY                107.222685
BURGLARY               106.403564
CYBERCRIME              49.625397
VEHICLE - STOLEN        48.879144
FIREARM OFFENSE         48.340980
ARSON                   48.305376
SHOPLIFTING             48.216842
VANDALISM               48.150259
FRAUD                   48.038618
COUNTERFEITING          47.739362
KIDNAPPING              47.707113
DRUG OFFENSE            47.692060
IDENTITY THEFT          47.489119
EXTORTION               47.457413
ILLEGAL POSSESSION      47.340980
ASSAULT                 47.282563
DOMESTIC VIOLENCE       47.111675
PUBLIC INTOXICATION     15.352000
TRAFFIC VIOLATION       14.793750
Name: Days Took To Close, dtype: float64

It is the average number of days it took to close cases in each crime description.

In [145]:
df.groupby('Crime Domain')['Victim Gender'].value_counts()
#.groupby(level=0).head(1)

Crime Domain      Victim Gender
Fire Accident     F                 2122
                  M                 1288
                  X                  415
Other Crime       F                12771
                  M                 7786
                  X                 2391
Traffic Fatality  F                 1074
                  M                  631
                  X                  210
Violent Crime     F                 6456
                  M                 3700
                  X                 1316
Name: count, dtype: int64

Here we can see the Victim's gender wise crime reported across each crime domain.
This was done to see which gender was targeted the most across each crime domain.

In [146]:
df.groupby('Crime Domain')['Crime Description'].value_counts()

Crime Domain      Crime Description  
Fire Accident     FIREARM OFFENSE        1931
                  ARSON                  1894
Other Crime       BURGLARY               1980
                  VANDALISM              1975
                  FRAUD                  1965
                  KIDNAPPING             1920
                  PUBLIC INTOXICATION    1912
                  HOMICIDE               1909
                  CYBERCRIME             1899
                  ILLEGAL POSSESSION     1895
                  DRUG OFFENSE           1890
                  EXTORTION              1873
                  COUNTERFEITING         1871
                  SHOPLIFTING            1859
Traffic Fatality  TRAFFIC VIOLATION      1915
Violent Crime     DOMESTIC VIOLENCE      1932
                  ROBBERY                1928
                  IDENTITY THEFT         1918
                  SEXUAL ASSAULT         1917
                  ASSAULT                1915
                  VEHICLE - STOLEN       1

Here we can the which crime domain contain which crime description and their counts.

In [147]:
df.head()

,Report Number,Date Reported,Date of Occurrence,Time of Occurrence,City,Crime Code,Crime Description,Victim Age,Victim Gender,Weapon Used,Crime Domain,Police Deployed,Case Closed,Date Case Closed,Days Took To Close
0,1,2020-01-02 00:00:00,01-01-2020 00:00,01-01-2020 01:11,Ahmedabad,576,IDENTITY THEFT,16,M,Blunt Object,Violent Crime,13,No,NaT,NaN
1,2,2020-01-01 19:00:00,01-01-2020 01:00,01-01-2020 06:26,Chennai,128,HOMICIDE,37,M,Poison,Other Crime,9,No,NaT,NaN
2,3,2020-01-02 05:00:00,01-01-2020 02:00,01-01-2020 14:30,Ludhiana,271,KIDNAPPING,48,F,Blunt Object,Other Crime,15,No,NaT,NaN
3,4,2020-01-01 05:00:00,01-01-2020 03:00,01-01-2020 14:46,Pune,170,BURGLARY,49,F,Firearm,Other Crime,1,Yes,2020-04-29 05:00:00,119.0
4,5,2020-01-01 21:00:00,01-01-2020 04:00,01-01-2020 16:51,Pune,421,VANDALISM,30,F,Other,Other Crime,18,Yes,2020-01-08 21:00:00,7.0


In [148]:
Safest_city = df['City'].value_counts().tail(1)
Most_unsafe_city = df['City'].value_counts().head(1)
Safest_city 


City
Rajkot    320
Name: count, dtype: int64

In [149]:
Most_unsafe_city

City
Delhi    5400
Name: count, dtype: int64

According to this Delhi is the most unsafe city and Rajkot is the safest city.

In [150]:
Top5_reasons = df.groupby('City')['Crime Description'].value_counts().loc['Delhi'].head(5)

Top 5 reasons for Delhi being the most unsafe city.

In [151]:
Top_crime_in_cities = df.groupby('City')['Crime Description'].value_counts().groupby(level=0).head(1).sort_values(ascending=False)
Top_crime_in_cities.head(10)

City       Crime Description  
Delhi      FIREARM OFFENSE        291
Mumbai     SEXUAL ASSAULT         243
Bangalore  PUBLIC INTOXICATION    198
Hyderabad  VEHICLE - STOLEN       165
Kolkata    FRAUD                  144
Chennai    ARSON                  138
Pune       VANDALISM              131
Ahmedabad  ASSAULT                104
Jaipur     COUNTERFEITING          87
Lucknow    TRAFFIC VIOLATION       83
Name: count, dtype: int64

Here we can see what top 10 cities are plagued with what crimes.

In [152]:
df['Reported_year'] = df['Date Reported'].dt.year

In [153]:
case_reported_per_year = df.groupby('Reported_year')['Reported_year'].value_counts()
case_reported_per_year

Reported_year
2020    8747
2021    8766
2022    8759
2023    8761
2024    5127
Name: count, dtype: int64

In [154]:
case_closed_per_year = df.groupby('Case Closed')['Reported_year'].value_counts().loc['Yes']
case_closed_per_year

Reported_year
2023    4467
2020    4413
2021    4333
2022    4308
2024    2541
Name: count, dtype: int64

Here we can clearly see the number of cases reported each years and number of cases closed each year.

In [155]:
case_reported_per_year.mean()

np.float64(8032.0)

In [156]:
case_closed_per_year.mean()

np.float64(4012.4)

Here we can the average number of cases reported each year is 8032 and average number of cases closed each year is 4012.

In [157]:
df['Time of Occurrence']= pd.to_datetime(df['Time of Occurrence'], format="%d-%m-%Y %H:%M")

In [158]:
df['Only_Time_of_Occurrence'] = df['Time of Occurrence'].dt.time

In [159]:
df['Only_Hours_of_Occurrence'] = df['Time of Occurrence'].dt.hour

In [160]:
bins=[0,4,8,12,16,20,24]
labels=['Midnight','Dawn','Morning','Afternoon','Evening','Night']
df['Time_period'] = pd.cut(df['Only_Hours_of_Occurrence'], bins=bins, labels=labels, right=False)

In [161]:
df.groupby('Time_period', observed=False)['Crime Description'].value_counts().groupby(level=0, observed=False).head(1)

Time_period  Crime Description
Midnight     SEXUAL ASSAULT       339
Dawn         VANDALISM            367
Morning      FRAUD                357
Afternoon    HOMICIDE             352
Evening      ROBBERY              350
Night        TRAFFIC VIOLATION    347
Name: count, dtype: int64

Here we can see what sort crimes are happening at what period of time.

In [162]:
df.head()

,Report Number,Date Reported,Date of Occurrence,Time of Occurrence,City,Crime Code,Crime Description,Victim Age,Victim Gender,Weapon Used,Crime Domain,Police Deployed,Case Closed,Date Case Closed,Days Took To Close,Reported_year,Only_Time_of_Occurrence,Only_Hours_of_Occurrence,Time_period
0,1,2020-01-02 00:00:00,01-01-2020 00:00,2020-01-01 01:11:00,Ahmedabad,576,IDENTITY THEFT,16,M,Blunt Object,Violent Crime,13,No,NaT,NaN,2020,01:11:00,1,Midnight
1,2,2020-01-01 19:00:00,01-01-2020 01:00,2020-01-01 06:26:00,Chennai,128,HOMICIDE,37,M,Poison,Other Crime,9,No,NaT,NaN,2020,06:26:00,6,Dawn
2,3,2020-01-02 05:00:00,01-01-2020 02:00,2020-01-01 14:30:00,Ludhiana,271,KIDNAPPING,48,F,Blunt Object,Other Crime,15,No,NaT,NaN,2020,14:30:00,14,Afternoon
3,4,2020-01-01 05:00:00,01-01-2020 03:00,2020-01-01 14:46:00,Pune,170,BURGLARY,49,F,Firearm,Other Crime,1,Yes,2020-04-29 05:00:00,119.0,2020,14:46:00,14,Afternoon
4,5,2020-01-01 21:00:00,01-01-2020 04:00,2020-01-01 16:51:00,Pune,421,VANDALISM,30,F,Other,Other Crime,18,Yes,2020-01-08 21:00:00,7.0,2020,16:51:00,16,Evening


In [163]:
ages= [0,18,20,50,80]
group= ['Under-18','Teenager','Adult','Elder']
df['Victim_Age_Group']= pd.cut(df['Victim Age'], bins=ages, labels=group, right=False)

In [164]:
df['Victim_Age_Group'].value_counts()


Victim_Age_Group
Adult       17331
Elder       17135
Under-18     4526
Teenager     1168
Name: count, dtype: int64

In [165]:
df['Victim_Age_Group'].value_counts(normalize=True)*100

Victim_Age_Group
Adult       43.154880
Elder       42.666833
Under-18    11.269920
Teenager     2.908367
Name: proportion, dtype: float64

Here we can see numbers as well as percentage of victim across different age groups.